File: backend/internal/service/events.go
package service

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/hujinrun/flowspace/internal/model"
	"github.com/hujinrun/flowspace/internal/storage"
)

var ErrInvalidEventProject = errors.New("invalid event project")

func GetEvents(ctx context.Context, store storage.Store, month string, page, pageSize int) ([]model.Event, int, error) {
	t, err := time.Parse("2006-01", month)
	if err != nil {
		return nil, 0, err
	}
	monthStart := t.Unix()
	monthEnd := t.AddDate(0, 1, 0).Unix()
	return store.Events().List(ctx, monthStart, monthEnd, page, pageSize)
}

func CreateEvent(ctx context.Context, store storage.Store, req *model.CreateEventRequest) (*model.Event, error) {
	projectID, err := validateEventProjectID(ctx, store, req.ProjectID)
	if err != nil {
		return nil, err
	}
	event := &model.Event{
		Title:     req.Title,
		StartTime: req.StartTime,
		EndTime:   req.EndTime,
		Location:  req.Location,
		Kind:      req.Kind,
		ProjectID: projectID,
	}
	if err := store.Events().Create(ctx, event); err != nil {
		return nil, err
	}
	return event, nil
}

func UpdateEvent(ctx context.Context, store storage.Store, id string, req *model.UpdateEventRequest) (*model.Event, error) {
	projectID, err := validateEventProjectID(ctx, store, req.ProjectID)
	if err != nil {
		return nil, err
	}
	updateReq := *req
	updateReq.ProjectID = projectID
	return store.Events().Update(ctx, id, &updateReq)
}

func DeleteEvent(ctx context.Context, store storage.Store, id string) error {
	return store.Events().Delete(ctx, id)
}

func validateEventProjectID(ctx context.Context, store storage.Store, projectID *string) (*string, error) {
	if projectID == nil {
		return nil, nil
	}
	trimmed := strings.TrimSpace(*projectID)
	if trimmed == "" {
		return &trimmed, nil
	}
	if _, err := store.Tasks().GetProjectByID(ctx, trimmed); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("%w: %s", ErrInvalidEventProject, trimmed)
		}
		return nil, err
	}
	return &trimmed, nil
}

File: frontend/src/api/events.ts
import { api } from './client'

export interface Event {
  id: string; title: string; start_time: number; end_time: number; location?: string; kind: string; note_id?: string; project_id?: string; project?: string; project_type?: string; created_at: number; updated_at: number
}

export async function getEvents(params: { month?: string; page?: number; page_size?: number }) {
  const res = await api.get<{ events: Event[] }>('/api/events', {
    month: params.month ?? '',
    page: String(params.page ?? 1),
    page_size: String(params.page_size ?? 50),
  })
  return { events: res.data.events, pagination: res.pagination! }
}

export async function createEvent(body: { title: string; start_time: number; end_time: number; location?: string; kind?: string; project_id?: string }) {
  const res = await api.post<{ event: Event }>('/api/events', body)
  return res.data.event
}

export async function updateEvent(id: string, body: Partial<Event>) {
  const res = await api.patch<{ event: Event }>(`/api/events/${id}`, body)
  return res.data.event
}

export async function deleteEvent(id: string) {
  await api.del(`/api/events/${id}`)
}
